helix-view/src/document.rs

```diff
@@ -162,6 +162,27 @@ impl Document {
         Ok(doc)
     }
 
+    pub fn reload(&mut self) -> Result<(), anyhow::Error> {
+        use std::{fs::File, io::BufReader};
+
+        let path = self
+            .path()
+            .ok_or_else(|| Error::msg("unable to reload document without path"))?;
+        let file = File::open(path).context(format!("unable to open {:?}", path))?;
+        let text = Rope::from_reader(BufReader::new(file))?;
+
+        self.changes = ChangeSet::new(&text);
+        self.history = Cell::new(History::default());
+        self.last_saved_revision = 0;
+        self.old_state = None;
+        self.text = text;
+        self.version = 0;
```

Comment:
I think you might need to keep this or LSP will get confused, not sure. I'm also not sure if the LSP detects external modifications or if we have to notify it of changes, this will need some testing.

Reply:
You also need to loop over `self.selections` and clamp all the ranges into valid ranges (imagine if I deleted half of the file then reloaded). You can use `selection.transform` for this

Reply:
Do we want to clamp the selections or just drop them? It's likely the contents have changed in ways that make the current selections wrong anyway, right?

Reply:
Regarding `self.version`. If the document hasn't changed, I guess it should stay the same. If the file on disk changed, do we want to increment `self.version` then? I'm not yet familiar with all its use cases except that it's used with LSP and incremented by `_apply`.

Reply:
Thinking about this further, perhaps the reload should be done using a `Transaction` allow you to undo the read like in vim?